lua/lsp/init.lua

```diff
@@ -115,41 +82,51 @@ function M.common_on_attach(client, bufnr)
   end
   lsp_highlight_document(client)
   add_lsp_buffer_keybindings(bufnr)
-  require("lsp.null-ls").setup(vim.bo.filetype)
 end
 
-function M.setup(lang)
-  local lsp_utils = require "lsp.utils"
-  local lsp = lvim.lang[lang].lsp
-  if (lsp.active ~= nil and not lsp.active) or lsp_utils.is_client_active(lsp.provider) then
+function M.setup_by_ft(lang, opts)
```

Comment:
It's annoying/wrong to have the opts connected to the language instead of the provider. If that's what you mean, then we're on the same page. 

Also, nearly all of those files in providers are gonna disappear as well, so we can freely change that struct.